### PR TITLE
Faults controller

### PR DESCRIPTION
**Pull request by AlexeyKozhevin**
- Add faults detector controller
- Add several layers for some attribute computation on GPU
- Remove aggregation functional
- Minor changes

### CODE REVIEW

File: seismiqb/src/layers.py [hunk 1]
```diff
@@ -29,33 +38,171 @@ def _hilbert(self, x):
...
+    fill_value : int, optional
+        Value to fill constant regions with std=0, by default 0.
+    """
+    def __init__(self, inputs=None, window=(1, 1, 100), padding='same', fill_value=0):
+        super().__init__()
+        self.window = window
+        self.padding = padding
+        self.fill_value = fill_value
+        self.ndim = inputs.ndim
+        self.kernel = torch.ones((1, 1, *window), dtype=torch.float32, requires_grad=False).to(inputs.device)
```
Comment (SergeyTsimfer) [inline seismiqb/src/layers.py:82]: `dtype=inputs.dtype`: can be changed by `autocast`

File: seismiqb/src/layers.py [hunk 2]
```diff
@@ -2,23 +2,32 @@
...
+from torch.cuda.amp import autocast
 
 from ..batchflow.models.torch import ResBlock
 
 
 
 class InstantaneousPhaseLayer(nn.Module):
-    """ Instantaneous phase comnputation. """
-    def __init__(self, inputs=None, continuous=False, enable=True, **kwargs):
+    """ Instantaneous phase computation.
```
Comment (SergeyTsimfer) [inline seismiqb/src/layers.py:13]: Add the fact that transforms are performed along the depth axis for all of your layers
Plus, add dev comments in the code for future self, as some parts of the code are cluttered and unintuitive :)

File: seismiqb/src/controllers/faults.py [hunk 3]
```diff
@@ -0,0 +1,708 @@
...
+            'output': 'sigmoid',
+            'itemwise': False
+        }
+    })
+
+    # Model (not controller) config
+    BASE_MODEL_CONFIG = {
+        'optimizer': {'name': 'Adam', 'lr': 0.01},
+        "decay": {'name': 'exp', 'gamma': 0.9, 'frequency': 100, 'last_iter': 2000},
+        'microbatch': C('microbatch'),
```
Comment (SergeyTsimfer) [inline seismiqb/src/controllers/faults.py:94]: Add defaults here as well: same as in controller_config

File: seismiqb/src/controllers/faults.py [hunk 4]
```diff
@@ -0,0 +1,708 @@
...
+            'attention': 'scse',
+            'phases': C('phase'),
+            'continuous': C('continuous_phase'),
+            'window': C('normalization_window'),
+            'normalization': C('normalization_layer')
+        },
+        'loss': C('loss')
+    }
+
+    UNET_CONFIG = {
```
Comment (SergeyTsimfer) [inline seismiqb/src/controllers/faults.py:109]: Do you really want to use `UNet` word in here?

File: seismiqb/src/controllers/faults.py [hunk 5]
```diff
@@ -0,0 +1,708 @@
...
+        },
+        'output': torch.sigmoid,
+        'common/activation': 'relu6',
+        'loss': C('loss')
+    }
+
+    # Train functional
+
+    def make_dataset(self, **kwargs):
+        """ Create an instance of :class:`.SeismicCubeset` with cubes and horizons. Arguments are from 'train'
```
Comment (SergeyTsimfer) [inline seismiqb/src/controllers/faults.py:152]: Arguments are from `dataset`?

File: seismiqb/src/controllers/faults.py [hunk 6]
```diff
@@ -0,0 +1,708 @@
...
+            ratios = {}
+
+            if len(dataset) > 0:
+                for i in range(len(dataset)):
+                    faults = dataset.labels[i]
+                    fault_area = sum([len(np.unique(faults[j].points)) for j in range(len(faults))])
+                    cube_area = np.prod(dataset.geometries[i].cube_shape)
+                    ratios[dataset.indices[i]] = fault_area / cube_area
+            else:
+                ratios[dataset.indices[0]] = 1
```
Comment (SergeyTsimfer) [inline seismiqb/src/controllers/faults.py:181]: Looks really suspicious

File: seismiqb/src/controllers/faults.py [hunk 7]
```diff
@@ -0,0 +1,708 @@
...
+
+    # Train functional
+
+    def make_dataset(self, **kwargs):
+        """ Create an instance of :class:`.SeismicCubeset` with cubes and horizons. Arguments are from 'train'
+        subconfig and kwargs. """
+        config = {**self.config['dataset'], **kwargs}
+        width = config['width']
+        label_dir = config['label_dir']
+        paths = [self.amplitudes_path(item) for item in config['train_cubes']]
```
Comment (SergeyTsimfer) [inline seismiqb/src/controllers/faults.py:157]: I would much rather prefer passing `train_cubes` directly, as this is the logical argument of making a dataset: 'create a dataset of these cubes for me'

File: seismiqb/src/controllers/faults.py [hunk 8]
```diff
@@ -0,0 +1,708 @@
...
+        transposed = config['transposed_cubes']
+        direction = {f'amplitudes_{cube}': 0 if cube not in transposed else 1 for cube in config['train_cubes']}
+
+        dataset = SeismicCubeset(index=paths)
+        dataset.load(label_dir=label_dir.format(width), labels_class=Fault, transform=True,
+                     direction=direction, verify=True)
+
+        return dataset
+
+    def make_sampler(self, dataset, ratios=None, threshold=0, **kwargs):
```
Comment (SergeyTsimfer) [inline seismiqb/src/controllers/faults.py:167]: Here, on the other hand, `ratios` and `threshold` are parameters, that should be taken from config

File: seismiqb/src/controllers/faults.py [hunk 9]
```diff
@@ -0,0 +1,708 @@
...
+                    fault_area = sum([len(np.unique(faults[j].points)) for j in range(len(faults))])
+                    cube_area = np.prod(dataset.geometries[i].cube_shape)
+                    ratios[dataset.indices[i]] = fault_area / cube_area
+            else:
+                ratios[dataset.indices[0]] = 1
+
+        weights = np.array([ratios[i] for i in dataset.indices])
+        weights /= weights.sum()
+        weights = weights.clip(max=0.3)
+        weights = weights.clip(min=0.1)
```
Comment (SergeyTsimfer) [inline seismiqb/src/controllers/faults.py:186]: Can be done in one clip

File: seismiqb/src/controllers/faults.py [hunk 10]
```diff
@@ -0,0 +1,708 @@
...
+    def make_notifier(self):
+        """ Make notifier. """
+        return Notifier(None, graphs=['loss_history',
+            {'source': B('images'), 'name': 'images', 'plot_function': self.custom_plotter},
+            {'source': B('masks'), 'name': 'masks', 'plot_function': self.custom_plotter},
+            {'source': B('predictions').astype('float32'), 'name': 'predictions', 'plot_function': self.custom_plotter},
+        ])
+
+    def custom_plotter(self, ax=None, container=None, **kwargs):
+        """ Zero-out center area of the image, change plot parameters. """
```
Comment (SergeyTsimfer) [inline seismiqb/src/controllers/faults.py:282]: ☹

File: seismiqb/src/controllers/faults.py [hunk 11]
```diff
@@ -0,0 +1,708 @@
...
+    def get_train_template(self, **kwargs):
+        """ Define the whole training procedure pipeline including data loading, augmentation and model training. """
+        return (
+            self.load_pipeline(create_masks=True, **kwargs) +
+            (self.augmentation_pipeline(**kwargs) if self.config['train/adjust'] else Pipeline()) +
+            self.train_pipeline(**kwargs)
+        )
+
+    def get_model_config(self, name):
+        """ Get model config depending model architecture. """
```
Comment (SergeyTsimfer) [inline seismiqb/src/controllers/faults.py:302]: Do you really need this function (and the next one)?

File: seismiqb/src/controllers/faults.py [hunk 12]
```diff
@@ -0,0 +1,708 @@
...
+            load pipeline for train (True) or for inference (False).
+        create_masks : bool, optional
+            create mask or not, by default True
+
+        Returns
+        -------
+        batchflow.Pipeline
+        """
+        if train and self.config['train/adjust']:
+            load_shape = F(adjust_shape)(
```
Comment (SergeyTsimfer) [inline seismiqb/src/controllers/faults.py:211]: Move shape adjusting to `Sampler` creation

File: seismiqb/src/controllers/faults.py [hunk 13]
```diff
@@ -0,0 +1,708 @@
...
+                C('crop_shape'), C('angle'), C('scale')[0]
+            )
+        else:
+            load_shape = C('crop_shape')
+
+        load_shape = F(np.array)(load_shape)
+        shape = {self.cube_name_from_alias(k): load_shape for k in self.config['dataset/train_cubes']}
```
Comment (SergeyTsimfer) [inline seismiqb/src/controllers/faults.py:218]: Use either cube original names or aliases everywhere

File: seismiqb/src/controllers/faults.py [hunk 14]
```diff
@@ -0,0 +1,708 @@
...
+                                 threshold=threshold, proportions=self.weights, **kwargs)
+
+        return sampler
+
+    def load_pipeline(self, train=True, create_masks=True):
+        """ Create loading pipeline common for train and inference stages.
+
+        Parameters
+        ----------
+        train : bool
```
Comment (SergeyTsimfer) [inline seismiqb/src/controllers/faults.py:201]: When the first parameter of `common train and inference pipeline` is `train bool`, it looks really sus

File: seismiqb/src/controllers/faults.py [hunk 15]
```diff
@@ -0,0 +1,708 @@
...
+            holes = binary_fill_holes(dilation, struct)
+            erosion = binary_erosion(holes, generate_binary_structure(2, 1))
+
+            processed_faults[slices] = binary_dilation(skeletonize(erosion, method='lee'))
+
+        return Fault.from_mask(processed_faults, prediction_cube, chunk_size=100, pbar=bar)
+
+# Train utils
+
+def adjust_shape(crop_shape, angle, scale):
```
Comment (SergeyTsimfer) [inline seismiqb/src/controllers/faults.py:613]: This (and all functions below) should be moved to a better place

**Review by SergeyTsimfer — COMMENTED** [review]
Really dislike the current usage of `Controller` config: it should contain only parameters of applying operations, and definitely not the paths to datasets and definitions of train/inference splits. Instead of this, methods should take 1-3 meaningful arguments and explicitly return created entities.

File: seismiqb/src/controllers/faults.py (hunk 7 above)
Comment (AlexeyKozhevin) [reply seismiqb/src/controllers/faults.py:157]: All that we can define in config must be defined in config.

File: seismiqb/src/controllers/faults.py [hunk 16]
```diff
@@ -0,0 +1,708 @@
...
+    def make_filename(self, prefix, orientation, ext):
+        """ Make filename for infered cube. """
+        return (prefix + datetime.now().strftime("%Y-%m-%d-%H-%M-%S") + '_{}.{}').format(orientation, ext)
+
+    # Inference utils
+
+    def skeletonize_faults(self, prediction, axis=0, threshold=0.1, bar=True):
+        """ Make faults from binary mask. """
+        prediction_cube = SeismicGeometry(prediction) if isinstance(prediction, str) else prediction
+        processed_faults = np.zeros(prediction_cube.cube_shape)
```
[the remark this answers is not in the document]
Comment (AlexeyKozhevin) [reply seismiqb/src/controllers/faults.py:595]: Now we have enough RAM so it is not a problem.
When we will find appropriate skeletonize procedure, we will make it more "eco-friendly"

File: seismiqb/src/cubeset.py [hunk 17]
```diff
@@ -560,8 +501,8 @@ def _compute_total_batches_in_all_chunks(self, idx, chunk_grid, chunk_shape, cro
 
     # Convenient loader
     def load(self, label_dir=None, filter_zeros=True, dst_labels='labels',
-             labels_class=Horizon, direction=None, **kwargs):
-        """ Load geometries and labels, stored on disk in a predefined format:
+             labels_class=None, direction=None, **kwargs):
```
Comment (SergeyTsimfer) [inline seismiqb/src/cubeset.py:504]: Revert changes

File: seismiqb/src/geometry/blosc.py [hunk 18]
```diff
@@ -217,21 +217,30 @@ def __repr__(self):
...
+            with self.zipfile.open(f'{self.key}/{key}', mode='w') as file:
+                self.dump(slide, file)
 
-        with self.zipfile.open(f'{self.key}/{key}', mode='w') as file:
-            self.dump(slide, file)
 
     def __getitem__(self, key):
         """ Load the file, named as the number of a slide. """
-        key = key if isinstance(key, int) else key[0]
+        key = key if isinstance(key, (int, slice)) else key[0]
```
Comment (SergeyTsimfer) [inline seismiqb/src/geometry/blosc.py:231]: You can preallocate `buffer` and put data in it

File: seismiqb/src/geometry/blosc.py [hunk 19]
```diff
@@ -217,21 +217,30 @@ def __repr__(self):
...
+                array.append(self[int(i)])
+            return np.stack(array, axis=0)
+
         for _ in range(self.RETRIES):
             # In a multi-processing setting, the ZipFile can be (somehow) closed from other process
             # We can mitigate that by re-opening the handler, if needed.
             try:
                 with self.zipfile.open(f'{self.key}/{key}', mode='r') as file:
                     slide = self.load(file)
-                break
```
Comment (SergeyTsimfer) [inline seismiqb/src/geometry/blosc.py:234]: Why?

File: seismiqb/src/metrics.py [hunk 20]
```diff
@@ -1090,3 +1090,59 @@ def blockwise(self, func, l=3, pbar=True, kernel=(5, 5), block_size=(1000, 1000)
...
+        res = array
+        axes = sorted(axes)
+        for i, axis in enumerate(axes):
+            res = res.sum(axis=axis-i)
+        return res
+
+    def make_shift(self, array, shift=20):
+        """ Make shifts for mask. """
+        result = np.zeros_like(array)
+        for i, _array in enumerate(array):
```
Comment (SergeyTsimfer) [inline seismiqb/src/metrics.py:1141]: Looks very like `np.roll`

File: seismiqb/src/utility_classes.py [hunk 21]
```diff
@@ -438,9 +438,9 @@ def _aggregate(self):
         elif self.type == 'numpy':
             self.counts[self.counts == 0] = 1
             if 'float' in self.dtype.__name__:
-                self.data /= self.counts
+                self.data[:] /= self.counts
```
Comment (SergeyTsimfer) [inline seismiqb/src/utility_classes.py:441]: Not needed anymore

File: seismiqb/src/controllers/faults.py [hunk 22]
```diff
@@ -0,0 +1,609 @@
...
+            'attention': 'scse',
+            'phases': C('phase'),
+            'continuous': C('continuous_phase'),
+            'window': C('normalization_window'),
+            'normalization': C('normalization_layer')
+        },
+        'loss': C('loss')
+    }
+
+    UNET_CONFIG = {
```
Comment (SergeyTsimfer) [inline seismiqb/src/controllers/faults.py:110]: You can actually move this to `best_practices.py`, which is a collection of various model configs (and probably should be renamed)

File: seismiqb/src/controllers/faults.py [hunk 23]
```diff
@@ -0,0 +1,609 @@
...
+
+    def make_sampler(self, dataset, **kwargs):
+        """ Create sampler for generating locations to train on. """
+        config = {**self.config['dataset'], **kwargs}
+        ratios = config['ratios']
+        threshold = config['threshold']
+        crop_shape = self.config['train']['crop_shape']
+
+        if self.config['train/augment'] and self.config['train/adjust']:
+            crop_shape = self.adjust_shape(crop_shape)
```
Comment (SergeyTsimfer) [inline seismiqb/src/controllers/faults.py:192]: Log this shape somewhere?

File: seismiqb/src/controllers/faults.py [hunk 24]
```diff
@@ -0,0 +1,609 @@
...
+        """ Make notifier. """
+        if self.config['train/visualize_crops']:
+            return Notifier(None, graphs=['loss_history',
+                {'source': B('images'), 'name': 'images', 'plot_function': self.custom_plotter},
+                {'source': B('masks'), 'name': 'masks', 'plot_function': self.custom_plotter},
+                {'source': B('prediction'), 'name': 'predictions', 'plot_function': self.custom_plotter},
+            ])
+        return super().make_notifier()
+
+    def custom_plotter(self, ax=None, container=None, **kwargs):
```
Comment (SergeyTsimfer) [inline seismiqb/src/controllers/faults.py:298]: Using `plot_image` could really help you here

File: seismiqb/src/controllers/faults.py [hunk 25]
```diff
@@ -0,0 +1,609 @@
...
+            cubes = {cube: (0, None, None, None) for cube in cubes}
+        for cube in cubes:
+            cubes[cube] = [cubes[cube]] if isinstance(cubes[cube][0], int) else cubes[cube]
+        return cubes
+
+    def make_accumulator(self, geometry, slices, crop_shape, strides, orientation=0, path=None):
+        """ Make grid and accumulator for inference. """
+        batch_size = self.config['inference']['inference_batch_size']
+        aggregation = self.config['inference']['aggregation']
+        name = 'cube_i'
```
Comment (SergeyTsimfer) [inline seismiqb/src/controllers/faults.py:395]: Make sure to update to `master`

File: seismiqb/src/controllers/faults.py [hunk 26]
```diff
@@ -0,0 +1,622 @@
...
+        return sampler
+
+    def load_train(self):
+        """ Create loading pipeline for train stages. """
+        return (Pipeline()
+            .make_locations(batch_size=C('batch_size'), generator=C('sampler'))
+            .load_cubes(dst='images', slicing=C('slicing'))
+            .create_masks(dst='masks')
+            .adaptive_reshape(src=['images', 'masks'])
+            .normalize(mode='q', itemwise=C('itemwise'), src='images')
```
Comment (SergeyTsimfer) [inline seismiqb/src/controllers/faults.py:234]: `q`-normalization with `itemwise=False` (which is the default value in your config) is not the best choice (and potentially breaks things)

**Review by SergeyTsimfer — APPROVED** [review]

**Review by SergeyTsimfer — APPROVED** [review]

**Review by SergeyTsimfer — APPROVED** [review]
